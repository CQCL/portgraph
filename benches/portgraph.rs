use criterion::{
    black_box, criterion_group, criterion_main, AxisScale, BenchmarkId, Criterion,
    PlotConfiguration,
};
use portgraph::{
    substitute::{BoundedSubgraph, OpenGraph, Rewrite},
    PortGraph, Weights,
};

/// Create line graph, connected with two parallel edges at each step.
///
/// o -2-> o -2-> o -2-> o -2-> o   ...
///
fn make_line_graph(size: usize) -> PortGraph {
    let mut graph = PortGraph::with_capacity(size, size * 2);
    let mut prev_node = graph.add_node(0, 2);

    for _ in 1..size {
        let new_node = graph.add_node(2, 2);
        graph.link_nodes(prev_node, 0, new_node, 0).unwrap();
        graph.link_nodes(prev_node, 1, new_node, 1).unwrap();
        prev_node = new_node;
    }

    graph
}

/// Create line graph, connected with two parallel edges at each step.
///
/// o -2-> o -2-> o -2-> o -2-> o   ...
///
#[allow(unused)]
fn make_weighted_line_graph(size: usize) -> (PortGraph, Weights<usize, isize>) {
    let mut g = PortGraph::with_capacity(size, size * 2);
    let mut weights = Weights::with_capacity(size, size * 2);
    let mut prev_node = g.add_node(0, 2);
    weights[prev_node] = 0;
    weights[g.output(prev_node, 0).unwrap()] = -1;
    weights[g.output(prev_node, 1).unwrap()] = -2;

    for i in 1..size {
        let w = i as isize;
        let new_node = g.add_node(2, 2);
        weights[new_node] = i;
        weights[g.input(new_node, 0).unwrap()] = w;
        weights[g.input(new_node, 1).unwrap()] = w + 1;
        weights[g.output(new_node, 0).unwrap()] = -w;
        weights[g.output(new_node, 1).unwrap()] = -w - 1;

        g.link_nodes(prev_node, 0, new_node, 0).unwrap();
        g.link_nodes(prev_node, 1, new_node, 1).unwrap();
        prev_node = new_node;
    }

    (g, weights)
}

/// Create an acyclic graph with `layer` layers and 2 nodes per layer, connected sequentially as follows.
///
/// o ---> o ---> o ---> o ---> o   ...
/// |                    |
/// v                    v
/// o ---> o ---> o ---> o ---> o   ...
///
fn make_two_track_dag(layers: usize) -> PortGraph {
    let mut graph = PortGraph::with_capacity(2 * layers, 6 * layers);
    if layers == 0 {
        return graph;
    } else if layers == 1 {
        graph.add_node(1, 1);
        graph.add_node(1, 1);
        return graph;
    }

    let mut prev_nodes = [graph.add_node(1, 1), graph.add_node(1, 1)];

    for layer in 1..layers {
        let new_nodes = [graph.add_node(1, 2), graph.add_node(2, 1)];
        graph.link_nodes(prev_nodes[0], 0, new_nodes[0], 0).unwrap();
        graph.link_nodes(prev_nodes[1], 0, new_nodes[1], 0).unwrap();
        // Add an edge connecting both nodes every third layer
        if layer % 3 == 0 {
            graph.link_nodes(new_nodes[0], 1, new_nodes[1], 1).unwrap();
        }
        prev_nodes = new_nodes;
    }

    graph
}

/// Remove one every five nodes from the graph.
fn remove_every_five(graph: &mut PortGraph) {
    let mut to_remove = Vec::new();
    for (i, v) in graph.nodes_iter().enumerate() {
        if i % 5 == 0 {
            to_remove.push(v);
        }
    }
    for node in to_remove {
        graph.remove_node(node);
    }
}

/// Remove nodes from the graph in an unordered way until it is empty.
fn remove_all_unordered(graph: &mut PortGraph) {
    while graph.node_count() > 5 {
        remove_every_five(graph);
    }
    // Remove all remaining nodes
    while graph.node_count() > 0 {
        graph.remove_node(graph.nodes_iter().next().unwrap());
    }
}

fn bench_make_portgraph(c: &mut Criterion) {
    let mut g = c.benchmark_group("graph creation");
    g.plot_config(PlotConfiguration::default().summary_scale(AxisScale::Logarithmic));

    for size in [100, 10_000, 1_000_000] {
        g.bench_with_input(
            BenchmarkId::new("make_line_graph", size),
            &size,
            |b, size| b.iter(|| black_box(make_line_graph(*size))),
        );
    }
    g.finish();
}

fn bench_clone_portgraph(c: &mut Criterion) {
    let mut g = c.benchmark_group("graph cloning");
    g.plot_config(PlotConfiguration::default().summary_scale(AxisScale::Logarithmic));

    for size in [100, 10_000, 1_000_000] {
        g.bench_with_input(
            BenchmarkId::new("clone_line_graph", size),
            &size,
            |b, size| {
                let graph = make_line_graph(*size);
                b.iter(|| black_box(graph.clone()))
            },
        );
    }
    g.finish();
}

fn bench_remove_unordered(c: &mut Criterion) {
    let mut g = c.benchmark_group("remove vertices unordered");
    g.plot_config(PlotConfiguration::default().summary_scale(AxisScale::Logarithmic));

    for size in [0, 100, 10_000] {
        g.bench_with_input(
            BenchmarkId::new("remove_vertices_unordered", size),
            &size,
            |b, size| {
                let graph = make_two_track_dag(*size);
                b.iter(|| black_box(remove_all_unordered(&mut graph.clone())))
            },
        );
    }
    g.finish();
}

fn generate_rewrite() -> (PortGraph, Rewrite) {
    let mut g = PortGraph::with_capacity(3, 2);

    let n0 = g.add_node(0, 2);
    let n1 = g.add_node(1, 1);
    let n2 = g.add_node(2, 0);

    g.link_nodes(n0, 0, n1, 0).unwrap();
    g.link_nodes(n1, 0, n2, 0).unwrap();

    let p0 = g.output(n0, 0).unwrap();
    let p1 = g.input(n2, 0).unwrap();

    let mut g2 = PortGraph::with_capacity(2, 1);
    // node to be inserted
    let n3 = g2.add_node(1, 1);
    let p2 = g2.input(n3, 0).unwrap();
    let p3 = g2.output(n3, 0).unwrap();

    let rewrite = Rewrite::new(
        BoundedSubgraph::new([n1].into_iter().collect(), vec![p0], vec![p1]),
        OpenGraph::new(g2, vec![p2], vec![p3]),
    );

    (g, rewrite)
}

fn bench_apply_rewrite(c: &mut Criterion) {
    let mut g = c.benchmark_group("run a rewrite operation");
    g.plot_config(PlotConfiguration::default().summary_scale(AxisScale::Logarithmic));

    g.bench_function("apply_rewrite", |b| {
        let (graph, rewrite) = generate_rewrite();
        b.iter(|| black_box(graph.clone().apply_rewrite(rewrite.clone())))
    });
    g.finish();
}

criterion_group!(
    benches,
    bench_make_portgraph,
    bench_clone_portgraph,
    bench_remove_unordered,
    bench_apply_rewrite,
);
criterion_main!(benches);
