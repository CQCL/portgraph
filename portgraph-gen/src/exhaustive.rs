use portgraph::PortOffset;
use std::{collections::BTreeSet, mem};

use crate::{
    builder::GraphBuilder,
    generator::{GraphData, GraphGenerator},
    Operation,
};

pub struct ExhaustiveGenerator;

type OperationLayer = Vec<(Operation, usize, usize)>;

impl ExhaustiveGenerator {
    pub fn new() -> Self {
        Self
    }

    fn generate_layers<'a>(
        &'a self,
        width: usize,
        max_ops: usize,
        last_indices: &'a BTreeSet<usize>,
    ) -> impl Iterator<Item = OperationLayer> + 'a {
        GenLayer {
            last_layer: Vec::new(),
            width,
            max_ops,
            last_indices,
        }
    }
}

struct GenLayer<'a> {
    last_layer: OperationLayer,
    width: usize,
    max_ops: usize,
    last_indices: &'a BTreeSet<usize>,
}

impl Iterator for GenLayer<'_> {
    type Item = OperationLayer;

    fn next(&mut self) -> Option<Self::Item> {
        let mut n_ops = self.last_layer.len();
        while let Some(last) = self.last_layer.pop() {
            let used_indices = all_used_indices(&self.last_layer);
            let next = next_op(&last, &used_indices, &self.last_indices, self.width);
            if let Some(next) = next {
                self.last_layer.push(next);
                break;
            }
        }

        if self.last_layer.is_empty() {
            n_ops += 1;
            if n_ops > self.max_ops {
                return None;
            }
        }

        while self.last_layer.len() < n_ops {
            let used_indices = all_used_indices(&self.last_layer);
            let start = self.last_layer.iter().map(|(_, i, _)| i).max();
            let start = start.map(|s| s + 1).unwrap_or(0);
            let fst_index = (start..self.width).find(|i| !used_indices.contains(i))?;
            // Incrementing this dummy op should give us the first valid op
            let dummy_op_zero = (Operation::Op3, fst_index, fst_index);
            let next = next_op(
                &dummy_op_zero,
                &used_indices,
                &self.last_indices,
                self.width,
            )?;
            self.last_layer.push(next);
        }

        Some(self.last_layer.clone())
    }
}

impl Operation {
    fn increment(&self) -> Option<Self> {
        match self {
            Operation::Op1 => Some(Operation::Op2),
            Operation::Op2 => Some(Operation::Op3),
            Operation::Op3 => None,
            _ => unreachable!(),
        }
    }
}

fn increment_index(
    mut index: usize,
    other_index: usize,
    used_indices: &BTreeSet<usize>,
    last_indices: &BTreeSet<usize>,
    max: usize,
) -> Option<usize> {
    while index + 1 < max {
        index += 1;
        if used_indices.contains(&index) {
            continue;
        }
        if !last_indices.contains(&other_index) && !last_indices.contains(&index) {
            continue;
        }
        return Some(index);
    }
    None
}

fn next_op(
    &(op, i, j): &(Operation, usize, usize),
    used_indices: &BTreeSet<usize>,
    last_indices: &BTreeSet<usize>,
    max_width: usize,
) -> Option<(Operation, usize, usize)> {
    if let Some(next_op) = op.increment() {
        return Some((next_op, i, j));
    }
    let next_op = Operation::Op1;

    let next_j = increment_index(j, i, &used_indices, last_indices, max_width);
    if let Some(next_j) = next_j {
        return Some((next_op, i, next_j));
    }

    let next_i = (i + 1..).find(|i| !used_indices.contains(i)).unwrap();
    let next_j = increment_index(next_i, next_i, &used_indices, last_indices, max_width);
    if let Some(next_j) = next_j {
        return Some((next_op, next_i, next_j));
    }

    None
}

fn all_used_indices(last_layer: &[(Operation, usize, usize)]) -> BTreeSet<usize> {
    last_layer.iter().flat_map(|&(_, i, j)| [i, j]).collect()
}

impl GraphGenerator for ExhaustiveGenerator {
    fn generate(&mut self, width: usize, depth: usize, num_graphs: usize) -> Vec<GraphData> {
        let mut results = Vec::with_capacity(num_graphs);

        let mut last_graphs = vec![(GraphBuilder::new(width), BTreeSet::from_iter(0..width))];
        while results.len() < num_graphs {
            let mut new_last_graphs = Vec::new();
            for (graph, last_indices) in mem::take(&mut last_graphs) {
                let max_ops = depth - graph.n_operations();
                for layer in self.generate_layers(width, max_ops, &last_indices) {
                    let mut graph_clone = graph.clone();
                    let mut last_indices = BTreeSet::new();
                    for (op, i, j) in layer {
                        graph_clone.add_operation(op, i, j);
                        last_indices.extend([i, j]);
                    }
                    results.push(graph_clone.clone().finish());
                    new_last_graphs.push((graph_clone, last_indices));
                }
            }
            if new_last_graphs.is_empty() {
                break;
            }
            last_graphs = mem::take(&mut new_last_graphs);
        }

        results.shrink_to_fit();
        results
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use super::*;
    use insta::assert_debug_snapshot;
    use portgraph::{
        algorithms::{toposort, TopoSort},
        Direction, LinkView, PortView,
    };

    #[test]
    fn test_generate_layers() {
        let generator = ExhaustiveGenerator::new();
        let last_indices = BTreeSet::from([0, 1, 2, 3]);
        let layers: Vec<_> = generator.generate_layers(4, 2, &last_indices).collect();
        assert_debug_snapshot!(layers);
    }

    #[test]
    fn test_generate_layers_2() {
        let generator = ExhaustiveGenerator::new();
        let last_indices = BTreeSet::from([0]);
        let layers: Vec<_> = generator.generate_layers(3, 1, &last_indices).collect();
        assert_debug_snapshot!(layers);
    }

    #[test]
    fn test_generate_layers_empty() {
        let generator = ExhaustiveGenerator::new();
        let last_indices = BTreeSet::new();
        let layers: Vec<_> = generator.generate_layers(2, 1, &last_indices).collect();
        assert!(layers.is_empty());
    }

    #[test]
    fn test_small_exhaustive_generation() {
        let mut generator = ExhaustiveGenerator::new();
        let graphs = generator.generate(2, 2, 10);
        let all_ops: Vec<_> = graphs.iter().map(|g| as_operations_vec(g)).collect();
        assert_debug_snapshot!(all_ops);
    }

    fn as_operations_vec(data: &GraphData) -> Vec<(Operation, usize, usize)> {
        let input_node = data
            .graph
            .nodes_iter()
            .find(|n| data.weights[n] == Operation::Input)
            .unwrap();
        let output_node = data
            .graph
            .nodes_iter()
            .find(|n| data.weights[n] == Operation::Output)
            .unwrap();
        let num_paths = data.graph.outputs(input_node).count();
        let mut port_to_qubit = BTreeMap::new();

        for path_index in 0..num_paths {
            let mut current_node = input_node;
            let mut current_offset = path_index;

            while current_node != output_node {
                let current_port = data.graph.output(current_node, current_offset).unwrap();
                let next_port = data.graph.port_link(current_port).unwrap();
                let next_node = data.graph.port_node(next_port).unwrap();
                port_to_qubit.insert(current_port, path_index);
                port_to_qubit.insert(next_port, path_index);

                current_node = next_node;
                current_offset = data.graph.port_offset(next_port).unwrap().index();
            }
        }

        let topo: TopoSort<_> = toposort(&data.graph, [input_node], Direction::Outgoing);
        topo.into_iter()
            .filter(|n| !matches!(data.weights[n], Operation::Input | Operation::Output))
            .map(|n| {
                (
                    data.weights[&n],
                    port_to_qubit[&data.graph.input(n, 0).unwrap()],
                    port_to_qubit[&data.graph.input(n, 1).unwrap()],
                )
            })
            .collect()
    }
}
