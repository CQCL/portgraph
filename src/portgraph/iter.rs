//! Iterator structures for a portgraph

use std::{
    iter::{Flatten, FusedIterator, Zip},
    num::NonZeroU32,
    ops::Range,
};

use crate::portgraph::{NodeEntry, PortEntry, PortGraph};
use crate::{NodeIndex, PortIndex, PortOffset};

/// Iterator over the ports of a node.
/// See [`PortGraph::inputs`], [`PortGraph::outputs`], and [`PortGraph::all_ports`].
#[derive(Debug, Clone)]
pub struct NodePorts {
    pub(super) index: NonZeroU32,
    pub(super) length: usize,
}

impl Iterator for NodePorts {
    type Item = PortIndex;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.length = self.length.checked_sub(1)?;
        let port = PortIndex(self.index);
        // can never saturate
        self.index = self.index.saturating_add(1);
        Some(port)
    }

    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.length = self.length.checked_sub(n + 1)?;
        let index = self.index.saturating_add(n as u32);
        self.index = index.saturating_add(1);
        Some(PortIndex(index))
    }

    fn count(self) -> usize {
        self.length
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.length, Some(self.length))
    }
}

impl ExactSizeIterator for NodePorts {
    fn len(&self) -> usize {
        self.length
    }
}

impl DoubleEndedIterator for NodePorts {
    #[inline]
    fn next_back(&mut self) -> Option<Self::Item> {
        self.length = self.length.checked_sub(1)?;
        let port = PortIndex(self.index.saturating_add(self.length as u32));
        Some(port)
    }
}

impl FusedIterator for NodePorts {}

impl Default for NodePorts {
    fn default() -> Self {
        Self {
            index: NonZeroU32::new(1).unwrap(),
            length: 0,
        }
    }
}

/// Iterator over the nodes of a graph, created by [`PortGraph::nodes_iter`].
#[derive(Clone)]
pub struct Nodes<'a> {
    pub(super) iter: std::iter::Enumerate<std::slice::Iter<'a, NodeEntry>>,
    pub(super) len: usize,
}

impl<'a> Iterator for Nodes<'a> {
    type Item = NodeIndex;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.iter.find_map(|(index, node_meta)| match node_meta {
            NodeEntry::Free(_) => None,
            NodeEntry::Node(_) => {
                self.len -= 1;
                Some(NodeIndex::new(index))
            }
        })
    }

    #[inline]
    fn count(self) -> usize {
        self.len
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.len, Some(self.len))
    }
}

impl<'a> ExactSizeIterator for Nodes<'a> {
    fn len(&self) -> usize {
        self.len
    }
}

impl<'a> DoubleEndedIterator for Nodes<'a> {
    fn next_back(&mut self) -> Option<Self::Item> {
        if self.len == 0 {
            return None;
        }

        while let Some((index, node_meta)) = self.iter.next_back() {
            if let NodeEntry::Node(_) = node_meta {
                self.len -= 1;
                return Some(NodeIndex::new(index));
            }
        }

        None
    }
}

impl<'a> FusedIterator for Nodes<'a> {}

/// Iterator over the ports of a graph, created by [`PortGraph::ports_iter`].
#[derive(Clone)]
pub struct Ports<'a> {
    pub(super) iter: std::iter::Enumerate<std::slice::Iter<'a, PortEntry>>,
    pub(super) len: usize,
}

impl<'a> Iterator for Ports<'a> {
    type Item = PortIndex;

    fn next(&mut self) -> Option<Self::Item> {
        if self.len == 0 {
            return None;
        }

        for (index, port_entry) in &mut self.iter {
            if let PortEntry::Port(_) = port_entry {
                self.len -= 1;
                return Some(PortIndex::new(index));
            }
        }

        None
    }

    fn count(self) -> usize {
        self.len
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.len, Some(self.len))
    }
}

/// Iterator over the port offsets of a node. See [`PortGraph::input_offsets`],
/// [`PortGraph::output_offsets`], and [`PortGraph::all_port_offsets`].
#[derive(Clone)]
pub struct NodePortOffsets {
    pub(super) incoming: Range<u16>,
    // Outgoing port offsets can go up to u16::MAX, hence the u32
    pub(super) outgoing: Range<u32>,
}

impl Iterator for NodePortOffsets {
    type Item = PortOffset;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(i) = self.incoming.next() {
            return Some(PortOffset::new_incoming(i as usize));
        }
        if let Some(i) = self.outgoing.next() {
            return Some(PortOffset::new_outgoing(i as usize));
        }
        None
    }

    fn count(self) -> usize {
        self.len()
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.len(), Some(self.len()))
    }
}

impl ExactSizeIterator for NodePortOffsets {
    fn len(&self) -> usize {
        self.incoming.len() + self.outgoing.len()
    }
}

impl DoubleEndedIterator for NodePortOffsets {
    fn next_back(&mut self) -> Option<Self::Item> {
        if let Some(i) = self.outgoing.next_back() {
            return Some(PortOffset::new_outgoing(i as usize));
        }
        if let Some(i) = self.incoming.next_back() {
            return Some(PortOffset::new_incoming(i as usize));
        }
        None
    }
}

impl FusedIterator for NodePortOffsets {}

/// Iterator over the links of a node, created by [`PortGraph::links`]. Returns
/// the port indices linked to each port, or `None` if the corresponding port is
/// not connected.
#[derive(Clone)]
pub struct NodeLinks<'a>(pub(super) std::slice::Iter<'a, Option<PortIndex>>);

impl<'a> Iterator for NodeLinks<'a> {
    type Item = Option<PortIndex>;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.0.next().copied()
    }

    #[inline]
    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.0.nth(n).copied()
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.0.size_hint()
    }
}

impl<'a> ExactSizeIterator for NodeLinks<'a> {
    #[inline]
    fn len(&self) -> usize {
        self.0.len()
    }
}

impl<'a> DoubleEndedIterator for NodeLinks<'a> {
    #[inline]
    fn next_back(&mut self) -> Option<Self::Item> {
        self.0.next_back().copied()
    }
}

impl<'a> FusedIterator for NodeLinks<'a> {}

/// Iterator over the neighbours of a node, created by
/// [`PortGraph::neighbours`]. May return duplicate entries if the graph has
/// multiple links between the same pair of nodes.
#[derive(Clone)]
pub struct Neighbours<'a> {
    graph: &'a PortGraph,
    linked_ports: Flatten<NodeLinks<'a>>,
}

impl<'a> Neighbours<'a> {
    /// Create a new iterator over the neighbours of a node, from an iterator
    /// over the links.
    pub fn from_node_links(graph: &'a PortGraph, links: NodeLinks<'a>) -> Self {
        Self {
            graph,
            linked_ports: links.flatten(),
        }
    }
}

impl<'a> Iterator for Neighbours<'a> {
    type Item = NodeIndex;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.linked_ports
            .next()
            .map(|port| self.graph.port_node(port).unwrap())
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.linked_ports.size_hint()
    }
}

impl<'a> DoubleEndedIterator for Neighbours<'a> {
    #[inline]
    fn next_back(&mut self) -> Option<Self::Item> {
        self.linked_ports
            .next_back()
            .map(|port| self.graph.port_node(port).unwrap())
    }
}

impl<'a> FusedIterator for Neighbours<'a> {}

/// Iterator over the links connecting two nodes, created by
/// [`PortGraph::get_connections`].
#[derive(Clone)]
pub struct NodeConnections<'a> {
    graph: &'a PortGraph,
    target: NodeIndex,
    port_links: Zip<NodePorts, NodeLinks<'a>>,
}

impl<'a> NodeConnections<'a> {
    /// Create a new iterator over the links connecting two nodes, from an
    /// iterator over the ports and links.
    pub fn new(
        graph: &'a PortGraph,
        target: NodeIndex,
        ports: NodePorts,
        links: NodeLinks<'a>,
    ) -> Self {
        Self {
            graph,
            target,
            port_links: ports.zip(links),
        }
    }

    /// Checks if we can yield an element from the `port_links` iterator.
    ///
    /// Receives `self.graph` and `self.target` as arguments to avoid borrowing issues.
    fn get_next(
        graph: &PortGraph,
        target: NodeIndex,
        next: (PortIndex, Option<PortIndex>),
    ) -> Option<(PortIndex, PortIndex)> {
        let (from, to) = next;
        let to = to?;
        if graph.port_node(to) == Some(target) {
            Some((from, to))
        } else {
            None
        }
    }
}

impl<'a> Iterator for NodeConnections<'a> {
    type Item = (PortIndex, PortIndex);

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.port_links
            .find_map(|next| Self::get_next(self.graph, self.target, next))
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.port_links.size_hint()
    }
}

impl<'a> DoubleEndedIterator for NodeConnections<'a> {
    #[inline]
    fn next_back(&mut self) -> Option<Self::Item> {
        self.port_links
            .try_rfold((), |_, next| {
                Self::get_next(self.graph, self.target, next).map_or(Ok(()), Err)
            })
            .err()
    }
}

impl<'a> FusedIterator for NodeConnections<'a> {}
