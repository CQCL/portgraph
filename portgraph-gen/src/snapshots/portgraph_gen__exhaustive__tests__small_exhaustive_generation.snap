---
source: portgraph-gen/src/exhaustive.rs
expression: all_ops
---
[
    [
        (
            Op1,
            0,
            1,
        ),
    ],
    [
        (
            Op2,
            0,
            1,
        ),
    ],
    [
        (
            Op3,
            0,
            1,
        ),
    ],
    [
        (
            Op1,
            0,
            1,
        ),
        (
            Op1,
            0,
            1,
        ),
    ],
    [
        (
            Op1,
            0,
            1,
        ),
        (
            Op2,
            0,
            1,
        ),
    ],
    [
        (
            Op1,
            0,
            1,
        ),
        (
            Op3,
            0,
            1,
        ),
    ],
    [
        (
            Op2,
            0,
            1,
        ),
        (
            Op1,
            0,
            1,
        ),
    ],
    [
        (
            Op2,
            0,
            1,
        ),
        (
            Op2,
            0,
            1,
        ),
    ],
    [
        (
            Op2,
            0,
            1,
        ),
        (
            Op3,
            0,
            1,
        ),
    ],
    [
        (
            Op3,
            0,
            1,
        ),
        (
            Op1,
            0,
            1,
        ),
    ],
    [
        (
            Op3,
            0,
            1,
        ),
        (
            Op2,
            0,
            1,
        ),
    ],
    [
        (
            Op3,
            0,
            1,
        ),
        (
            Op3,
            0,
            1,
        ),
    ],
]
